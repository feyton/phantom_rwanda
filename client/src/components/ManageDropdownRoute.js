import React, { useRef, useState } from 'react';
import { useNavigate } from 'react-router-dom';
import { toast } from 'react-toastify';
import Swal from 'sweetalert2';
import { axiosBase as axios } from '../utils/Api.js';
import CheckRole from '../utils/CheckRoles.js';
import { ButtonLoading } from './Button.js';

function ManageDropdownRoute({ row }) {
  const selectRef = useRef();
  const { id } = row;
  const navigate = useNavigate();
  const [loading, setloading] = useState(false);
  const handleEdit = () => {
    navigate(`route/edit/${id}`, {
      state: {
        route: row
      }
    });
  };
  const handleView = () => {
    navigate(`/dashboard/modal/routes/view/${id}`, {
      state: {
        route: row
      }
    });
  };
  const handleDelete = async () => {
    const results = await Swal.fire({
      title: 'Are you sure?',
      html: `Route <b>${row.origin}-${row.destination}</b> will be deleted.`,
      text: 'This Route will be deleted!',
      icon: 'warning',
      showCancelButton: true,
      confirmButtonColor: '#3085d6',
      cancelButtonColor: '#d33',
      confirmButtonText: 'Yes, delete!'
    });
    if (results.isConfirmed) {
      setloading(true);
      try {
        await axios.delete(`/routes/${id}`);
        toast('Route Deleted successfully', { type: 'success' });
        setloading(false);
        if (selectRef.current) {
          selectRef.current.innerHTML = `<p class="font-raleway mx-auto font-bold text-red">Deleted</p>`;
        }
      } catch (error) {
        toast(error.message, { type: 'error' });
        setloading(false);
      }
    }
  };

  const handleChange = (e) => {
    switch (e.target.value) {
      case 'view':
        handleView();
        break;
      case 'edit':
        handleEdit();
        break;
      case 'deleteRoute':
        handleDelete();
        break;
      default:
        break;
    }
  };

  return (
    <div className="z-10" ref={selectRef}>
      {loading ? (
        <ButtonLoading name="Sending" />
      ) : (
        <select
          role="select"
          onChange={handleChange}
          name=""
          id="select"
          className="py-1 px-3 font-rale font-bold bg-transparent border rounded-sm"
        >
          <option hidden> Manage </option>
          <option value="view"> View </option>
          <CheckRole
            children={
              <option className="cursor-pointer" value="edit">
                {' '}
                Edit{' '}
              </option>
            }
            role={['operator']}
          />
          <CheckRole
            children={
              <option className="cursor-pointer" value="deleteRoute">
                {' '}
                Delete{' '}
              </option>
            }
            role={['operator']}
          />{' '}
        </select>
      )}
    </div>
  );
}

export default ManageDropdownRoute;
