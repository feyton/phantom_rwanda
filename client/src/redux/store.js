import { configureStore } from '@reduxjs/toolkit';
import { authReducer } from './reducers/authReducer.js';
import { busMovementReducer } from './reducers/busMovementReducer.js';
import { counterReducer } from './reducers/counterReducer.js';
import { dashboardReducer } from './reducers/dashboardReducer.js';
import { managementReducer } from './reducers/managementReducer.js';
import { profileReducer } from './reducers/profileReducer.js';
import { simulateReducer } from './reducers/simulateReducer.js';
import { socketReducer } from './reducers/socketReducer.js';
import { updateProfileReducer } from './reducers/updateProfileReducer.js';

export const store = configureStore({
  reducer: {
    counter: counterReducer,
    profileUpdates: updateProfileReducer,
    profileInfo: profileReducer,
    auth: authReducer,
    management: managementReducer,
    dashboard: dashboardReducer,
    socket: socketReducer,
    simulate: simulateReducer,
    movement: busMovementReducer
  }
});
