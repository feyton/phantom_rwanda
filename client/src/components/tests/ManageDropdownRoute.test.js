import { fireEvent, render, screen } from '@testing-library/react';
import { mount } from 'enzyme';
import { enableFetchMocks } from 'jest-fetch-mock';
import React from 'react';
import * as redux from 'react-redux';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router-dom';
import renderer from 'react-test-renderer';
import { store } from '../../redux/store.js';
import ManageDropdownRoute from '../ManageDropdownRoute.js';

let spyOnUseSelector;
enableFetchMocks();

describe('ManageDropdownRoute component', () => {
  spyOnUseSelector = jest.spyOn(redux, 'useSelector');
  spyOnUseSelector.mockReturnValue({
    authenticated: true,
    user: {
      first_name: 'Fabrice',
      role: 'admin'
    }
  });

  const row = {
    routes: {
      destination1: 'Nyabugogo',
      destination2: 'DownTown',
      distance: '5km',
      id: 15,
      total_buses: '100'
    }
  };
  it('should render the ManageDropdownRoute component', () => {
    const elem = renderer
      .create(
        <Provider store={store}>
          <MemoryRouter>
            <ManageDropdownRoute row={row} />
          </MemoryRouter>
        </Provider>
      )
      .toJSON();
    expect(elem).toMatchSnapshot();
  });

  it('should render the CheckRole component', () => {
    const wrapper = mount(
      <Provider store={store}>
        <MemoryRouter>
          <ManageDropdownRoute row={row} />
        </MemoryRouter>
      </Provider>
    );

    expect(wrapper.find('CheckRole').length).toEqual(2);
  });

  it('should Edit a Route', async () => {
    const handleEditMock = jest.fn();
    const elem = render(
      <Provider store={store}>
        <MemoryRouter>
          <ManageDropdownRoute row={row} handleEdit={handleEditMock} />
        </MemoryRouter>
      </Provider>
    );
    fireEvent.change(screen.getByTestId('select'), {
      target: { value: 'edit' }
    });
  });

  it('should View a Route', async () => {
    const handleViewMock = jest.fn();
    const elem = render(
      <Provider store={store}>
        <MemoryRouter>
          <ManageDropdownRoute row={row} handleView={handleViewMock} />
        </MemoryRouter>
      </Provider>
    );
    fireEvent.change(await screen.getByRole('select'), {
      target: { value: 'view' }
    });
  });

  it('should Delete selected Route', async () => {
    const handleDeleteMock = jest.fn();
    const elem = render(
      <Provider store={store}>
        <MemoryRouter>
          <ManageDropdownRoute row={row} handleDelete={handleDeleteMock} />
        </MemoryRouter>
      </Provider>
    );
    fireEvent.change(screen.getByTestId('select'), {
      target: { value: 'delete' }
    });

    expect(elem).toMatchSnapshot();
  });

  const mockedUsedNavigate = jest.fn();
  jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useNavigate: () => mockedUsedNavigate
  }));
  const mockedUsedLocation = jest.fn();
  jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useLocation: () => mockedUsedLocation
  }));

  it('should not Fetch data from the database', async () => {
    const mockNavigator = jest.fn();
    jest.mock('react-router-dom', () => ({ useNavigate: () => mockNavigator }));

    global.fetch.mockResponseOnce(
      JSON.stringify({
        routes: {
          destination1: 'Kabuga',
          destination2: 'Nyabugogo',
          distance: '20km',
          total_buses: '100',
          id: 13
        }
      })
    );

    // expect(global.fetch.mock.calls.length).toEqual(1)
    expect(global.fetch).toHaveBeenCalledTimes(0);
  });
});
