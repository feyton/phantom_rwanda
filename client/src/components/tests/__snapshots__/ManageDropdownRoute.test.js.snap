// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ManageDropdownRoute component should Delete selected Route 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="z-10"
      >
        <select
          class="py-1 px-3 font-rale font-bold bg-transparent border rounded-sm"
          id="select"
          name=""
          role="select"
        >
          <option
            hidden=""
          >
             Manage 
          </option>
          <option
            value="view"
          >
             View 
          </option>
          <option
            class="cursor-pointer"
            value="edit"
          >
             Edit 
          </option>
          <option
            class="cursor-pointer"
            value="deleteRoute"
          >
             Delete 
          </option>
           
        </select>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="z-10"
    >
      <select
        class="py-1 px-3 font-rale font-bold bg-transparent border rounded-sm"
        id="select"
        name=""
        role="select"
      >
        <option
          hidden=""
        >
           Manage 
        </option>
        <option
          value="view"
        >
           View 
        </option>
        <option
          class="cursor-pointer"
          value="edit"
        >
           Edit 
        </option>
        <option
          class="cursor-pointer"
          value="deleteRoute"
        >
           Delete 
        </option>
         
      </select>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`ManageDropdownRoute component should render the ManageDropdownRoute component 1`] = `
<div
  className="z-10"
>
  <select
    className="py-1 px-3 font-rale font-bold bg-transparent border rounded-sm"
    id="select"
    name=""
    onChange={[Function]}
    role="select"
  >
    <option
      hidden={true}
    >
       Manage 
    </option>
    <option
      value="view"
    >
       View 
    </option>
    <option
      className="cursor-pointer"
      value="edit"
    >
       
      Edit
       
    </option>
    <option
      className="cursor-pointer"
      value="deleteRoute"
    >
       
      Delete
       
    </option>
     
  </select>
</div>
`;
